### DISCUSSION TITLE
[Proposal] New `ref` convention for returning references

### DISCUSSION THREAD

**Discussion by nmsmith** [modular/modular#2874] [Ideas]
Hi everyone,

@lattner and I have developed an alternative "auto-deref" proposal to the one that Chris posted [a few weeks ago](https://github.com/modularml/mojo/discussions/2594). The new idea is to make auto-dereferencing a result convention, rather than a feature of the `Reference` type.

The proposal can be found [here](https://github.com/modularml/mojo/blob/nightly/proposals/ref-convention.md).

We'd love to hear feedback from the Mojo community. You can share your thoughts below.

**Comment by mzaks** [comment]
Great Proposal! I have two things which popped up in my mind. 

1. I think calling the Reference type just a Pointer doesn’t do it justice, or rather might cause false associations with users coming from other languages. The whole SafePointer vs UnsafePointer is IMHO very judgmental 😄. As it doesn’t translate the why, it implicitly tells you don’t use the second one always use the first one. In my head TrackedPointer makes more sense.

2. Regarding return type, how about infixing the fact that the return is not by value but rather a location e.g. `fn foo() -> Int@<origin>`:

**Reply by nmsmith** [reply]
Thanks for the feedback! I'll respond to each of your points separately:
1. I guess I'm not the best person to comment on the naming scheme for pointers in Mojo. Mojo will end up with several "pointer-like" types and obviously, they can't all be called "pointer". I expect that `Reference` (whatever its name ends up being) will be the "canonical" pointer type. It will be the thing that the average Mojo user reaches for when they want to point to a variable located somewhere else. In that light, it makes sense that the name is short and simple. `UnsafePointer` is something that only low-level "hackers" need to know about, so it deserves a more sophisticated name. 🙂
2. The issue with `Int@<origin>` is that it looks like a fancy _**type**_, and the new conventions are not types. If we were to use that syntax, I'd be worried that a lot of learners would attempt to declare types like `List[Int@<origin>]` and wonder why that doesn't work. Ideally then, the syntax we end up with would start with the origin/lifetime and a _name_ for the argument/result (if applicable), and end with the type. And bonus points if the type is delimited by `:`.

**Comment by dmitry-salin** [comment]
### Rename `Reference` to `Pointer`

When `ref` is explicitly promoted to `Pointer` with a call to its constructor, to people coming from other languages it may look like conversion from safe reference to unsafe pointer. I think `ref` could be renamed to `ptr` for consistency. Thus, `Pointer` and `ptr` can be documented as safe type and safe convention respectively.

### Syntax for the convention

I think the square bracket syntax is more consistent because `ref` convention is generic over lifetime and address space. This way, parameterization syntax is the same for types and convention, and no additional keywords are required.
```
fn foo(...) -> ref[lifetime, addr_space] T: ...
```
vs
```
fn foo(...) -> ref from lifetime, addr_space: T: ...
```

### What about lifetime elision rules?

Looks like this is a syntactic sugar for `ref[_] self` with elided lifetime:
```
fn __add__(borrow self, rhs: MyInt) -> MyInt: ...
fn __iadd__(mut self, rhs: MyInt): ...
```

This also looks like example of lifetime elision:
```
for (ref elt) in mutableList:
    elt = 42
```

These rules could potentially apply in many situations. In this example, the returned `ref` has `__lifetime_of(self)`:
```
fn __getitem__(ref self, index: Int) -> ref Self.ElementType:
```

The LSP may have a configuration option to always display elided lifetimes as inlay hints. This way they can still be explicit to the user.

**Reply by nmsmith** [reply]
`ptr` is definitely an option worth considering.

Regarding lifetime elision, I expect Mojo will do less of that than Rust. IMO it makes sense for arguments (the lifetime of each argument is an _implicit_ parameter by default, as you say), but not so much for return types, because the lifetime associated with a returned reference affects how long it is usable for. Obscuring this information might do more harm than good.

In general, I personally believe the best way to make lifetimes accessible to Mojo users is to make them _simpler_: a more intuitive syntax, an easier to explain model, and less restrictive borrowing rules. We should be able to achieve all of this.

**Reply by dmitry-salin** [reply]
Since Mojo is another approach to lifetimes, it doesn't necessarily have to copy the Rust model. But some common things are inevitable and should be learned by users. The lifetime elision concept is just one option that could be used more widely at any time later as it does not interfere with this proposal. `__lifetime_of(self)` is such a common pattern that it will be everywhere, so time will tell whether it is better for users to write it or just read it as inlay hints.

**Comment by yinonburgansky** [comment]
I greatly appreciate the enhanced ergonomics provided by the new ref return type for users interacting with the `Reference` type.

## questions

I do have some questions and concerns I'd like to ask:

1. How can we prevent the "infection" of user code when calling functions that return:

- other reference types like `Arc[T]` / `Mutex[T]`.
- nested reference types like `List[Reference[T]]`, `List[Arc[T]]`.
- `SomeStruct` with a field of type `Reference[T]` / `Arc[T]`.

2. The decision whether to auto-deref or not, is shifted from the *user* to the *library author*.
Since the concept of *restoring* the Reference type on the user side, applies only to the *builtin Reference* type, but not to other reference types like `Arc` / `Mutex` / `RWLock` due to information lost. Therefore, the auto-deref feature remains unique to the `builtin ref` type, managed by the compiler and inaccessible from the user side. unlike other builtin types which are accessible from the user side.
Library authors aiming to enhance user experience for those unfamiliar with References must provide 2 function signatures: one for returning types like `T` / `List[T]`, and another for returning special Reference types like `Arc[T]` / `Mutex[T]` / `List[Arc[T]]` / `List[Mutex[T]]`.

3. Limiting `borrowed` / `inout` to the builtin ref type, rather than allowing it to work with other reference types, such as sending a `RWLock` to `inout` argument, will `acquire` and `release` the lock upon `read` / `write`. This enables end-user functions to work with any reference type *transparently* without ever mentioning `Reference`, working seamlessly in single and multi threaded processing. However, this proposed convention may impose a higher learning curve for getting good performance in multi-threaded applications.

4. Using `__getitem__` for setting/mutating the reference, deviates from Python convention of `__setitem__`. While understandable due to other mutations like `+=`, the name is a bit misleading, and it lacks "hooks" for `acquiring` and `releasing` resources for reference types like `Mutex` / `RWLock`.

5. IMHO working with references remains verbose as constant dereferencing is required (search for `[]` in the `stdlib`). Implementing proper auto-deref support, as proposed earlier would significantly reduce verbosity, since *reading* from a reference is more common than *writing* to it. This would make the common case more concise and easier to read, especially with multi-threaded reference variants prevalent in multi-threaded processing. 

----

## Some thoughts

I believe the previous proposal allowing auto-deref for user-defined types holds more potential.
From what I understand from the discussions, a key missing component is a mechanism to prevent auto-deref when manual handling of the actual reference is necessary.

I am *hopeful* that a solution to this issue exists.

Disclaimer:  
As primarily a Python user, I am still in the process of learning about `mojo`, `lifetimes`, and the `borrow-checker`. Therefore, I may be overlooking or misunderstanding certain aspects of how they function; I apologize for any such oversight. During my exploration of references and examination of the `stdlib` code, I couldn't help but notice the widespread use of `[]`, which often distracts from the essence of the code. Personally, I advocate for its sparing usage, reserving it for situations where it is absolutely necessary. The following are some ideas I've come up with (disregarding keyword renaming and syntax discussions):

## preventing auto-dereferencing

Introducing "no-auto-deref variables" with a special syntax can effectively *disable* auto-dereferencing when working with these variables:

```mojo
var x = 1
var &a = Reference(Reference(x))  # No auto-deref since `a` is declared with `&` (consider postfix `&` or other syntax).
var b = a  # Reference[Reference[Int]], since `a` is declared with `&` it doesn't auto-deref.
var c = b  # auto-deref to Int, equivalent to a[][]
var d: Reference[Int] = b  # Control decay by providing an explicit type annotation, equivalent to a[]

var e = List(b, d)  # List[Int] since `b` and `d` auto-deref.
var f = List[Reference[Int]](b, d)  # List[Reference[Int]] control auto-deref with type annotation.
var g = List(a)  # List[Reference[Reference[Int]]] since `a` doesn't auto-deref.
```

system programmers who wish to work with raw references without auto-deref, can prefix variable *declarations* with &, similar to `C`/`C++`.

```mojo
var lst: List[Reference[Reference[Int]]]

for x in lst: ...  # `x` is Int
for &x in lst: ...  # `x` is Reference[Reference[Reference[Int]]]

var [a, *b] = lst  # `a` is Int, `b` is Iterable[Int]
var [&a, *&b] = lst  # `a` is Reference[Reference[Int]], `b` is Iterable[Reference[Reference[Int]]]
```

the rules:
1. When the `LHS` variable is declared with the no-auto-deref symbol `&` - the `RHS` expression won't auto-deref.
2. Any variable declared with no-auto-deref symbol `&` will not auto-deref, regardless of where it is used and require manual dereferencing with `[]`.
3. Explicit type annotation can be utilized to control auto-deref.

### complexity 

If simplicity remains a concern, and we aim for even simpler rules, we can eliminate rule 3 entirely, as it's not essential for complete functionality. This effectively reduces the rules to:

There are 2 variable types:
- "Regular" variable declaration which will *always* auto-deref all the way down.
- "Reference" variables which will *never* auto-deref and require manual dereferencing with `[]`.

This essentially shifts the decision of auto-dereferencing from the `callee-signature` to the `caller-"signature"` defined by the prefix `&` variable declaration. 
The primary distinction lies in not discriminating between `builtin-ref` and `user-reference`, maintaining consistency with existing builtin types thus far.

----

## integration with `inout` and `borrowed`

Enabling `inout` and `borrowed` to work with other reference types, such as `Mutex` and `RWLock` can be achieved through a mechanism like so:

```mojo
trait Referable[T, lifetime]:
    fn __enter__(self) -> Referable[T, lifetime]: ... # called when borrowing the reference

    fn __exit__(self):
        pass  # By default, no action when the reference is returned after borrowing.


struct RWLock[T, lifetime](Referable[T, lifetime]):
    fn __enter__(self) -> Referable[T, lifetime]:
        @parameter
        if self.lifetime.is_mutable:
            # handle write... acquire lock
            return self.unsafe_ptr
        else:
            # handle read...

    fn __exit__(self):
        # release lock

```

Any type that confirms to the `Referable` trait can be declared via `borrowed` or `inout`.

```mojo
borrowed x: T = some Referable[T, AnyLifetime]
inout y: T = some Referable[T, MutableLifetime]
```

When `reading` or `writing` from a `borrowed` or `inout` variable, a context manager will be opened to access the reference.

```mojo
var x = 1
var a: Reference[Reference[Int]] = x  # Implicit conversion and allowed to auto-deref

borrowed b = a  # `b` acts like Int, auto-deref as far as possible
inout c = a  # `c` acts like Int, auto-deref as far as possible

c = 2  # Mutate Reference[Int], by calling the context manager methods __enter__ and __exit__ before and after the mutation.
print(x, b, c, a[][])  # prints 2 2 2 2
```

Using with no-auto-deref syntax:

```mojo
var x = 1
var y = 2
var &a = Reference(Reference(x))
var &b = Reference(Reference(y))

inout &c = a  # Acts like Reference[Int]
inout &d = d
c, d = d, c # Swap the references, mutate Reference[Reference[Int]], calls __enter__ and __exit__
print(x, y, a[][], b[][], c[], d[]) # prints 1 2 2 1 2 1 
```

This allows users, unaware of reference types, to work with nested reference types:

```mojo
var lst: List[Reference[Int]]

for inout x in lst:
    x += 1  # Mutate Reference[Int], *not* the list item; user doesn't need to know if it's an Int or Reference[Int].

for inout &y in lst:
    x = ... some Reference[Int] # Mutate Reference[Reference[Int]] the list item.
```

Additionally, functions with `inout`/`borrowed` args will work seamlessly with *any* `Referable` Type.

```mojo
# User function unaware of all reference types and deref syntax
def do_something(inout x, borrowed y): ...


# Send it any Referable Type
var x: RWLock[Int]
var y: Arc[Int]
do_something(x, y) # User function only knows about Int, and automatically calls `__enter__`/`__exit__` without knowing exact types, acquiring and releasing the lock.
```

In the above code, the lock is acquired automatically without the user needing to be aware of it. In an ideal scenario without deadlock concerns, the function can simply auto-wait for the lock, allowing for a *gradual* introduction and adaptation to multi threaded code. This facilitates sharing the same user function for single-threaded with Reference type or multi-threaded with `Arc`/`Mutex`/`RWLock` types.

When a user requires more performance or is more advanced, they can always modify the function signature to specifically request the lock, such as for batching mutations with a single lock acquire-release. However, it's beneficial that users can learn about it gradually.

When sending non-`Referable` types to functions accepting `borrowed`/`inout` arguments, they will auto-promote to the built-in `Reference` type by default, similar to current behavior:

```mojo
def do_something(inout x, borrowed y): ...

var x = 8
var y = 9
do_something(x, y)  # auto-promote builtin Reference, equivalent to:
do_something(Reference(x), Reference(y))
```

Regarding expressions returning `Referable` types being on the `LHS`, there may be ambiguity with `__setattr__`. One potential solution is to allow it only with the `inout` keyword, as follows:

```mojo
var x = 8
inout Reference(x) = 9  
# allowed only with the `inout` keyword as a shorthand for:
inout temp = Reference(x)
temp = 9


struct SomeStruct:
    var a: Reference[Int]

var y: SomeStruct = ...
inout y.a = 8  # Mutate Reference[Int] since used with keyword inout
y.a = Reference... # Mutate the actual field a with __setattr__

var &z = Reference(x)
inout (z) = 10  # Specify an expression explicitly to prevent confusion with declaration and allow for potential shadowing in the future.  
inout z = 10  # Error, `z` is already declared.
```

**Reply by nmsmith** [reply]
Thanks for your feedback. If I'm understanding you correctly, you're mostly concerned about how the `ref` convention would generalize to "concurrent references", which manifests as `Arc<Mutex<T>>` in Rust.

I think the right thing to do here would be to first figure out what "concurrent references" would look like in Mojo, and more generally, what idiomatic concurrent programming looks like in Mojo. We might not end up with the same model as Rust. In particular, mutexes are a problematic concurrency primitive, owing to the risk of deadlock, so it would be good if the average Mojo user could work with something higher-level. (Deadlock is an "aliasing problem", similar to the problem solved by lifetimes.)

Once we understand all of this, we're more likely to know how it interacts with Mojo's argument conventions.

All that said, I'm merely an external contributor, so I don't have much insight into Mojo's future concurrency model. Perhaps somebody from Modular can provide a more informed response.

**Reply by Brian-M-J** [reply]
Mojo can dodge all these issues by adopting a concurrency model that integrates well with value semantics (so no need for references and lifetimes) and doesn't require users to work with locks directly (so no need to worry about `Arc`, `Mutex` etc.). An example of this is [senders and receivers](https://github.com/modularml/mojo/pull/1445#issuecomment-1873298549).

**Comment by rd4com** [comment]
Good job! about names, just an idea: SafeReference for Reference and UnsafeReference for UnsafePointer.

That way, maybe less confusion as to why Reference can be made with an UnsafePointer and so on ?

But this is just an idea, excellent job ! :fire:

**Comment by melodyogonna** [comment]
+1 on renaming Reference to "SafePointer" rather than just "Pointer". Due to C and C++, pointers, by themselves, generally have negative connotations. it'll be great to convey at a glance (to both low-level and high-level programmers) that this pointer is different and a lot safer.

**Comment by midrange22** [comment]
This is a great proposal! 

I want to voice my support for the proposed keyword alternatives for `inout`, `borrowed`, and `owned`.
The proposed new keywords `init`, `borrow`, `mut`, and `consume` makes things much clearer when reading Mojo code, in my opinion.
The example shown in the proposal has sold me on the idea:

```
## Proposed
struct MyInt:
   fn __init__(init self, value: Int): ...

   fn __add__(borrow self, rhs: MyInt) -> MyInt: ...

   fn __iadd__(mut self, rhs: MyInt): ...

   fn __del__(consume self): ...
```

**Comment by taylorpool** [comment]
By virtue of the fact that Mojo already positions itself as a safe language, I believe that prepending `Safe` onto types such as `Reference` or `Pointer` would be redundant.

**Reply by melodyogonna** [reply]
I believe that adding "safe" to the name will instantly clarify things for many people. Otherwise, they might be confused by the abundance of unsafe operations at first glance.

As I mentioned earlier, when people think of pointers, the first thing that comes to mind is C/C++ pointers, which are unsafe. Even raw pointers in Rust are considered unsafe.

**Comment by martinvuyk** [comment]
I think we should have less names.

I'm not familiar with concrete implementations of languages so I might be completely off.

But to me it looks like we only really need 3 concepts:
- read only reference
- mutable reference
- reference whose lifetime will end (be consumed) 

Why not have 3 types that describe exactly that and have auto-deref? 
And have Reference be the generic type that has to be manually dereferenced.

```mojo
struct MyInt:
   fn __init__(self: MutRef[Self], value: Int): ...

   fn __add__(self: Ref[Self], rhs: Self) -> Self: ...

   fn __iadd__(self: MutRef[Self], rhs: Self): ...

   fn map(self: ConsumeRef[Self]) -> Self: ...

   fn __del__(self: ConsumeRef[Self]): ...
```

And simpler to reason about lifetime notation (Lifetime should be able to be constructed from the Ref subtypes)

```mojo
fn some_func(arg1: Ref[Int], arg2: MutRef[Int]) -> Ref[String]:
    if arg1 > arg2:
        return Reference[String, False, Lifetime(arg1)]("bigger")
    return Reference[String, False, Lifetime(arg1)]("smaller")


fn some_func2(arg1: Ref[Int], arg2: MutRef[Int]) -> Mutref[String]:
    if arg1 > arg2:
        return Reference[String, True, Lifetime(arg1)]("bigger")
    return Reference[String, True, Lifetime(arg1)]("smaller")


fn some_func3[
    a: Lifetime, b: Lifetime
](arg1: Reference[Int, False, a], arg2: Reference[Int, True, b]) -> Ref[String]:
    if arg1[] > arg2[]:
        return Reference[String, False, a]("bigger")
    return Reference[String, False, a]("smaller")
```

**Reply by remorses** [reply]
But `ConsumeRef` is not a ref, it's a value. And `Ref` is the default for arguments but not for result types, that's what's confusing imo.

**Reply by nmsmith** [reply]
@martinvuyk In Mojo, mutability is associated with lifetimes. A reference can be used to mutate its pointee if and only if its lifetime is mutable. This is why it makes sense to have just one `Reference` struct (plus the `ref` convention for auto-deref), rather than a different struct for each mutability.

Also, `consume` is used to declare that a function deinitializes a variable. It isn't directly related to references.

Regarding auto-deref, the problems with having structs auto-deref themselves was discussed in the thread for [the former proposal](https://github.com/modularml/mojo/discussions/2594), and it was also mentioned in the new proposal.

**Comment by soraros** [comment]
Don't have time to go into details, here are some potential thinking points:
- Not being able to get auto-derefed values from `Tuple` could be a bigger problem than people think. It's a special (and maybe the simplest) case of destructing (think `ref` pattern in Rust). It will be appear everywhere, once we have more pattern matching built out. `ref` as of now can be think of as customising the semantics of the binding form: `var Ctor(name, ...) = f(...)`.
- We could/should write `ref[...]` instead of `ref [...]`, "parametric convention marker" if you will. Let's don't start with cute syntax using `from` etc.
- Let's hold off the keyword renaming etc to later please. Not until we have more serious use of refs at least.

**Reply by lattner** [reply]
I agree on delaying the keyword renaming fyi, it's more to provide a vision than to suggest we do it all at once.

**Comment by dderooy** [comment]
Thanks for the great write up! A quick thought: `alter` instead of `mut`.
All together then: `init`, `borrow`, `alter`, `consume`.

**Reply by dderooy** [reply]
`modify` or `mod` would be a good key name too and goes well with modular 😉

**Comment by lattner** [comment]
Status update: the first three steps of this proposal are all implemented, and should roll out in the next nightly. I'm very happy with how they fit together.

**Comment by Brian-M-J** [comment]
I'm just hoping Mojo leans harder towards value semantics so I don't have to deal with all this, or at least won't have to deal with it in the overwhelming majority of situations.

**Comment by ivellapillil** [comment]
```
struct MyInt:
   fn __init__(init self, value: Int): ...

```

Could it mean that I can implement any method as a "constructor"/factory method and `__init__` just becomes a convention?

E.g.

```
struct MyInt:
   @staticmethod
   fn new(init self, value: Int) -> ref Self: ...

# Usage

MyInt.new(1)

```

Or will we restrict the `init` calling convention to `__init__`?

P.S. I understand that `__init__` is treated specially, but once we have the calling conventions, it need not be? `init` becomes the indicator for a constructor like operation (as it indicates that we pass an uninitialized thing and by end of the function we expect it to be initialized and optionally(?) returned.

**Comment by forfudan** [comment]
Thanks For the proposal.

I support `init`, `read`, `mut`. Not very sure about `consume`. reasons are as follows.

- `mut` would be more friendly for a Rust user.
- I prefer `read` because it literally indicates absolute immutable. Instead, `borrow` can still mean that you can make some changes to it, like a mutable ref.
- About `consume`, I am not sure whether it is the best name. The value is only consumed when a `^` sigil is attached. Otherwise, it is copied. So the default action is copy instead of consume. `own` seems better because it leaves users to decide whether it is copied or consumed (with the `^` sigil).

**Reply by martinvuyk** [reply]
+1 on `mut`
+1 on `read`, it is clearer than `borrow` as you said
+1 on `own` instead of `consume`. I think `consume` shouldn't exist. `owned` as is is already clear in that you do whatever you want to the value inside your function. Copies (as far as I know) are implicit when calling the function only if you use the passed value after calling it. I think I like the idea of reducing it to 3 letters.

I personally do not like `init`, I think it can be just `mut` in the `__init__` func IMO

**Reply by nmsmith** [reply]
`consume` is semantically correct within the *body* of the function. An argument marked as `consume` *must* be consumed within the function body using the ^ sigil, or by invoking the destructor (which if invoked explicitly, will always use the ^ sigil as well).

So `consume` explains what the function does to the variable it is given.

*What* variable the function is given is an independent concern, and this is where copyability comes into play, as you both mention. If a variable has an implicitly-copyable type, then passing it to a function as `consume`, without writing the ^ sigil, will create a copy of that variable, and then pass the copy into the function. As mentioned earlier, the function will then consume the copy, using the ^ sigil.

I agree the copying behaviour is a bit subtle, but I’m not sure if there’s a simpler way to explain all of this. I don’t think `own` adds clarity, because it’s not clear what ownership means in this case. In C++ and Mojo, when we talk about ownership we are normally referring to ownership of MEMORY, and a function receiving a variable as `consume` crucially does *not* own the variable’s memory. The memory remains allocated after the function returns, and it still belongs to its original owner. The only difference is that the memory has become uninitialized. The `own` keyword doesn’t make clear that this deinitialization is going to occur.

The best way to understand `consume` is that it’s the semantic inverse of `init`. So the keywords would ideally be antonyms.

In short: naming is hard. 🥲

**Reply by forfudan** [reply]
I agree with you. The function indeed "consumes" the argument in any cases, either consume the variable or a copy of it. The confusion mainly comes from the actions outside the function call, i.e., from the implicit copy of the variable before it is passed to the function. Let's say,

```
fn ownit(owned x: String) -> String:
    var y = x
    return y

var x = str("something")
y = ownit(x)
```

Normally, when people see this code, they would think that the `ownit(x)` gets the memory of `x` (like in Rust). But actually no, `ownit(x)` actually gets `x.copy()`. Only if a `^` is added, i.e., `ownit(x^)`, the function takes the memory of `x`.

Although within the function block, the argument `x` is always consumed, however, due to this implicit copying, the de facto behavior of the function deviates from what is indicated by the keyword `consume`.

Semantically speaking, this is what happened:

|                               | without `^`                 | with `^`        |
| ----------------------------- | --------------------------- | --------------- |
| before pass into function | Make a copy of `x`          | Noting          |
| within function                   | **Consume** the copy of `x` | **Consume** `x` |

Yeah, within the function, consume always happens, so `consume` is definitely most proper verb.

But from a user's view, this is what they sees:

|                           | without `^`                                    | with `^`                                        |
| ------------------------- | ---------------------------------------------- | ----------------------------------------------- |
| before `var y = ownit(x)` | `x` owns memeory `m1`                          | `x` owns memeory `m1`                           |
| after `var y = ownit(x)`  | `x` owns memeory `m1`<br> `y` **owns** memory `m2` | `x` is uninitialized  <br> `y` **owns** memory `m1` |

A user *without knowing the implicit copy before `x` is passed into the function* would intuitively finds that:
- The original `x` is consumed and uninitialized if and only if a `^` is attached. (So calling `^` a consume operator is very accurate.)
- After the function call, the function owns something, not necessarily the original value (memory), but at least it owns something.

Thus, `consume` becomes conditional (on `^`) and `own` becomes absolute.

Well, it is all about from which angle we view the things. But I think the second table would be more intuitive to most of the new users, because they cannot directly see how compiler works.

So the question is: are there some verbs that can simultaneously and properly describes the two tables?

Yes, I got one, although may be surprising to you, but it is logical:

`dup` or `duplicate`

```
fn ownit(dup x: String) -> String:
    var y = x
    return y

var x = str("something")
y = ownit(x^)
```

Why? Because by default, a copy of `x` is made before it is passed into the function. The consume operator `^` can override this rule and directly pass `x` into the function (`x^` says, "do not make a copy of me!"). This is exactly the task of `^`: the original `x` is consumed if and only if a `^` is attached.

**Reply by martinvuyk** [reply]
> `dup` or `duplicate`

Interesting, how about `copy` ? I think it's easier to explain that it copies the value or consumes if the `^` sigil is used

**Reply by ivellapillil** [reply]
The key point is that whatever is getting passed, is always _consumed_ by the function. The function itself does not care if it is a copy or the actual value. It consumes. Basically like fire - it does not care what you throw into it, it just consumes it.

The copy is happening _before_ the function is being called. So the copy term belongs to the _caller_, not the function. Now, ^ can be thought of "move" and whatever does not have ^ is potentially getting copied (I say potentially as compiler might optimize away the copy). And this is an important point to note. Copy does not _always_ occur. It is the decision of the compiler to copy or not. So using `copy` or `dup` as keyword would be more confusing (at least for me) as it would not be true for many cases.

For me `consumed` is the most appropriate, as it actually tells what is happening _inside_ the method. 

From an API point of view, the function is a blackbox, with only the arguments giving us a hint what it does. So an arguments that tells me it will consume is more helpful than an argument that tells me the caller _may_ make a copy. Since I am the caller, I already know what my code is doing. I don't know what the API is doing (without access to doc/source code of the function)...

**Reply by forfudan** [reply]
> The copy is happening before the function is being called. So the copy term belongs to the caller, not the function.

Although the `copy` happens before the function is being called, this `copy` step is completely hidden from the eyes of programmers. What I did is only one thing: call the function pass `x` into it. If mojo compiler implicitly makes a copy before the function, I am completely not aware of that. It is not "what I see is what I get".

I like the idea behind duck type. Analogically, *if calling the function behaves as making a copy, then it is a copy.*

I would raise a "funny but logical" example. Assume that `^` now is a negative operator. Considering the following:

```mojo
fn negative(x: Int) -> Int:
    return -x

var x = 1
# Implicitly insert: x = -x unless there is a ^ sigil
print(negative(x)) # You get 1.
print(negative(x^)) # You get -1.
```

Why? Because the compiler implicitly insert `x = -x` before calling the function unless there is a `^` sigil. Thus, although the function is explicitly named as `negative`, it actually behaved as doing nothing.

The content in the function body is doing nothing wrong, i.e., taking negative value. But the user will find calling the function very unintuitive, i.e., `negative(1)` returns 1.

The example above is quite similar to the `consume` case. We defined the function with `consume`, but we what we see is a `copy`.

So here are two extra solutions:

- To make assignment a move, and make `^` a copy operator. That is pretty like Rust, but not consistent with the philosophy of Mojo. I do not think it is good.
- To find a more neutral word as a replacement of keyword `owned`. Using `copy` is not consistent with what is defined in the function per se. Using `consume` is not consistent what is seen when calling the function without `^`. Maybe a neutral keyword like `get` or `value` can soft-pedal the confusion, making it purely a symbol.

**Reply by ivellapillil** [reply]
But calling it `copy` will also not convey the fact that sometimes compiler does _not_ copy even if we do not use ^. (For e.g. if the developer expected `__copyinit__` to be called, they would be surprised that (if the variable is not used later on after the function call), it is moved to the function and the `__copyinit__` will not get called. So I don't think we have a solution where we can make something explicit, when the compiler itself makes the decisions implicitely.

What we are sure is that the variable is consumed within the function. So we can make that part explicit.

**Reply by martinvuyk** [reply]
I think `duplicate` i.e. `dup` as you said earlier aligns with what you're saying. I agree it is not clear yet in an "intuitive" way what `owned` does. How about `clone` ? It would either be a copy or the "real" one when used with the `^` sigil.

**Reply by ivellapillil** [reply]
Duplicate and clone are basically synonyms of copy... My main point is that copy does not happen always and therefore would be misleading.

**Reply by forfudan** [reply]
> I think `duplicate` i.e. `dup` as you said earlier aligns with what you're saying. I agree it is not clear yet in an "intuitive" way what `owned` does. How about `clone` ? It would either be a copy or the "real" one when used with the `^` sigil.

I agree with @ivellapillil that `dup` or `copy` are not consistent with what has been defined in the function body, although they may be consistent with what users see.

Maybe a mid-way is good (using a verb that are not that direct), or completely remove `owned` keyword? The argument is passed as `owned` by default. When you want to pass a immutable ref, use `borrowed`.

**Reply by ivellapillil** [reply]
> Maybe a mid-way is good (using a verb that are not that direct), or completely remove `owned` keyword? The argument is passed as `owned` by default. When you want to pass a immutable ref, use `borrowed`.

I think one of the insights in Mojo has been that most of the code is `borrowed` by default and we can be more ergonomic if we make it the default. It also allows some nice effects, that by default the compiler does not have to make potentially expensive copies.  To prevent unnecessary copies, the developers have to be careful to mark their arguments with `borrowed`, which add to their cognitive load. Quite often developers do the "minimum" to get the code working, so if `borrowed` is not impacting the working of the function, they will leave it out. The nice thing about `borrowed` as default is that it forces them to use `owned` or `inout` if they want to mutate the argument. This means that they are marking an argument because it has direct impact to their function. Since `borrowed` is the more efficient compared to the others, making it the default actually makes the code faster by default - even if the developer did not think of it.

**Reply by forfudan** [reply]
Yes, it makes sense to make immutable reference default option.

But in that way, we still have to think about assigning a nicer, and less confusing keyword for `owned`.  :D

**Comment by lattner** [comment]
This is all basically working now and the discussion converged, thank you for driving this @nmsmith !!
